mrc-1427 mint impact graphs spec

hillalex opened the pull request:
Ok, so this is much more abstracted now. The graph component is just a generic wrapper around plotly and is totally agnostic about what type of graph it is, the api data determines everything.

I've written this up into a possible API schema here: https://github.com/mrc-ide/mintr/pull/1

hillalex committed:
try to extract spec for data from graphs and tabl

hillalex committed:
Merge branch 'master' into mrc-1443

hillalex committed:
extract api interface

hillalex committed:
abstract further

hillalex committed:
merged master

hillalex committed:
abstract further

hillalex commented on mint/Impact.vue:
```diff
@@ -0,0 +1,55 @@
+        computed: {
+            selectedSettings() {
+                const result: any = {};
+                settings.map(s => {result[s.id] = s.value});
```
The interface for the settings is essentially very similar to the dynamic form so that could maybe be re-used here - just need to pass a form values style object to the graph and table components for them to do what they want with.

hillalex committed:
typo

hillalex commented on mint/ImpactTable.vue:
```diff
@@ -0,0 +1,42 @@
+<template>
```
Really this could just be called `DynamicTable` or similar - there's nothing actually particular to the impact data here.

hillalex commented on mint/fakeAPIData.ts:
```diff
@@ -0,0 +1,326 @@
+    };
+
+export const settings = [
+    {
```
I'm thinking these settings could use the dynamic form schema, aka the same as this: https://github.com/mrc-ide/hintr/blob/master/inst/schema/ModelRunOptions.schema.json

EmmaLRussell replied:
Excellent idea. :+1:

hillalex committed:
use real scale fake data

EmmaLRussell commented:
Do I have to do something cleverer than just build and open the file now? Getting `Fetch API cannot load file:///home/emma/dev/vue-prototypes/mint/public/data.json. URL scheme must be "http" or "https" for CORS request.`

EmmaLRussell commented on mint/PlotlyGraph.vue:
```diff
@@ -0,0 +1,67 @@
+                return true;
+            },
+            getRow(id: string) {
+                return this.dataSet.find((row: any) => row["intervention"] == id && this.filterBySettings(row));
```
"intervention" seems quite specific here, should this be a prop? Maybe it's universal in mint, but might not be in other apps if we re-use. Maybe a global const rather than a prop?

hillalex replied:
Yep agree, I think it can be added to the graph metadata easily enough

EmmaLRussell commented on mint/PlotlyGraph.vue:
```diff
@@ -0,0 +1,67 @@
+            Plotly
+        },
+        computed: {
+            dataSeries() {
```
Sorry, I don't quite understand what's going on here. This is making the distinction between barchart type data and line graph type data based on its shape? What's the difference between data and dataset? Maybe a couple of brief comments would be useful.

hillalex replied:
Agree, these should be renamed

hillalex replied:
(data and dataSet are actually metadata and data, respectively, nothing to do with chart type)

hillalex committed:
use real scale data, use composition api to reuse code

hillalex commented:
> Do I have to do something cleverer than just build and open the file now? Getting `Fetch API cannot load file:///home/emma/dev/vue-prototypes/mint/public/data.json. URL scheme must be "http" or "https" for CORS request.`

I use the IDE - right click on the html file and select "run in browser"

hillalex committed:
add types and make naming clearer

hillalex commented:
Sorry should possibly have done the Vue3 stuff on a separate branch, but see what you think. Vue3 is not released yet (it's in beta, likely release within a couple months) but they have created a plugin fir Vue2 which adds the new composition API, which is what I'm using here.

Have also added some types and changed the names of those confusingly named props - hopefully that makes it clearer what's going on...

hillalex commented on mint/PlotlyGraph.vue:
```diff
@@ -0,0 +1,81 @@
+                        // all values are given explicitly
+                        return d
+                    }
+                    if (d.id && props.metadata.format == "wide") {
```
In fact, this would be even clearer split out into 2 components now that the metadata is at graph level rather than series level - let me try that.

hillalex committed:
split long and wide

hillalex commented:
Ok and I've also split the graph component up based on whether the data format is wide vs long format

EmmaLRussell reviewed (approved):
The Vue3 syntax seems pretty nice!
